### PR TITLE
Feature/1623 implement public components for usage in module settings components

### PR DESCRIPTION
**Pull request by jqlee85**
## Summary

Addresses issue #1623 

## Relevant technical choices

 * Split functionality from `assets/js/components/settings/settings-module.js` into:
   * `DefaultModuleSettings`
   * `ModuleSettings`
   * `ModuleSettingsHeader`
   * `ModuleSettingsBody`
   * `ModuleSettingsFooter`
* Created an extra `displayMode`, `saving`, with an extra selector `isSavingSettings` as a shortcut for checking that state.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by eugene-manuilov** [comment]
@googlebot I consent.

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 1]
```diff
@@ -240,6 +249,60 @@ const baseActions = {
...
+	/**
+	 * Sets settings component for a module.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string} slug                 Module slug.
+	 * @param {WPElement} settingsComponent React component to render the settings panel. Default is the DefaultModuleSettings component.
+	 * @return {Object} Redux-style action.
+	 */
+	*setSettingsComponent( slug, settingsComponent ) {
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/datastore/modules.js:262]: We shouldn't have this action, the settings component should be provided as part of `registerModule`.

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 2]
```diff
@@ -292,6 +356,47 @@ const baseReducer = ( state, { type, payload } ) => {
 				},
 			};
 		}
+		case SET_SETTINGS_COMPONENT: {
+			return {
+				...state,
+			};
+		}
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/datastore/modules.js:363]: See above, this should be removed.

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 3]
```diff
@@ -292,6 +356,47 @@ const baseReducer = ( state, { type, payload } ) => {
...
+		case SET_SETTINGS_DISPLAY_MODE: {
+			const { slug, status } = payload;
+			const { modules: existingModules } = state;
+
+			// Clone object so as not to override previous object's values below.
+			const updatedModules = JSON.parse( JSON.stringify( existingModules ) );
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/datastore/modules.js:369]: I think we can avoid using `JSON` functions here. How about we create the object as `updatedModules = { ...existingModules }` and then below, wherever we modify a module we use `updatedModules[ slug ] = { ...updatedModules[ slug ], displayMode: '...' }`.

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 4]
```diff
@@ -355,21 +460,22 @@ const baseSelectors = {
...
-				} else {
-					moduleWithComponent.settingsComponent = DefaultModuleSettings;
-				}
+
+			// If there is a settingsComponent that was passed use it, otherwise set to the default.
+			if ( ModuleComponents[ registryKey ] && ModuleComponents[ registryKey ][ module.slug ] ) {
+				moduleWithComponent.settingsComponent = ModuleComponents[ registryKey ][ module.slug ];
+			} else {
+				moduleWithComponent.settingsComponent = undefined;
 			}
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/datastore/modules.js:474]: This shouldn't be changed from what it was before. This is related to my other comment: `DefaultModuleSettings` should have exactly the purpose indicated by its name - it should be the default module settings component. The component implemented further above should rather be `SimpleModuleSettings` or something like that.

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 5]
```diff
@@ -403,6 +509,80 @@ const baseSelectors = {
 		return modules[ slug ];
 	} ),
 
+	/**
+	 * Gets the display mode for module's settings.
+	 *
+	 * Returns one of: "closed", "view", "edit", or "locked".
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/datastore/modules.js:515]: Per this PR, can also be "saving". Let's rely on constants though, as mentioned in my other comment.

File: assets/js/modules/analytics/index.js [hunk 6]
```diff
@@ -16,28 +16,30 @@
...
 	compose( fillFilterWithComponent )( AnalyticsSetup )
 );
 
-addFilter(
-	'googlesitekit.ModuleSettingsDetails-analytics',
-	'googlesitekit.AnalyticsModuleSettings',
-	compose( fillFilterWithComponent )( AnalyticsSettings )
-);
+domReady( () => {
+	dispatch( CORE_MODULES ).setSettingsComponent( 'analytics', AnalyticsSettings );
```
Comment (felixarntz) [inline assets/js/modules/analytics/index.js:44]: See other comment, let's use `registerModule`.

File: includes/Core/Modules/Modules.php [hunk 7]
```diff
@@ -804,13 +804,17 @@ private function update_module_settings( Module $module, array $data ) {
...
-			'homepage'     => $module->homepage,
-			'internal'     => $module->internal,
-			'order'        => $module->order,
+			'name'         => isset( $module_info['name'] ) ? $module_info['name'] : $module->name,
+			'description'  => isset( $module_info['description'] ) ? $module_info['description'] : $module->description,
+			'homepage'     => isset( $module_info['homepage'] ) ? $module_info['homepage'] : $module->homepage,
+			'internal'     => isset( $module_info['internal'] ) && $module_info['internal'],
+			'order'        => isset( $module_info['sort'] ) ? $module_info['sort'] : $module->order,
+			'provides'     => isset( $module_info['provides'] ) ? $module_info['provides'] : array(),
+			'autoActivate' => isset( $module_info['autoActivate'] ) && $module_info['autoActivate'],
```
Comment (felixarntz) [inline includes/Core/Modules/Modules.php:817]: Oh now I see you already changed this. Generally +1, however:
* `autoActivate` should become `forceActive` and rely on `$module->force_active`.
* `provides` should not be part of this, let's handle that one purely on the client. In the long term it should be removed from the server.

Also, the new property needs to be covered in the `get_module_schema()` method too.

File: assets/js/modules/analytics/components/settings/SettingsEdit.js [hunk 8]
```diff
@@ -63,39 +61,6 @@ export default function SettingsEdit() {
...
-		);
-
-		return () => {
-			removeFilter(
-				'googlekit.SettingsConfirmed',
-				'googlekit.AnalyticsSettingsConfirmed',
-			);
-		};
-	}, [] );
-
```
Comment (felixarntz) [inline assets/js/modules/analytics/components/settings/SettingsEdit.js:98]: See my overarching PR review comment: We cannot remove legacy fixes like this until we fully migrate away from the old architecture.

File: assets/js/googlesitekit/modules/components/DefaultModuleSettings.js [hunk 9]
```diff
@@ -16,8 +16,91 @@
...
+const { useSelect } = Data;
+
+function DefaultModuleSettings( props ) {
+	const {
+		slug,
+		onView,
+		onEdit,
+		onSave,
+		canSave,
+		canDisconnect,
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/googlesitekit/modules/components/DefaultModuleSettings.js]: Updated.

File: assets/js/googlesitekit/modules/components/ModuleSettingsFooter.js [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+		}
+	}, [] );
+
+	const {
+		autoActivate,
+		homepage,
+		name,
+		connected,
+		provides,
+	} = module;
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/googlesitekit/modules/components/ModuleSettingsFooter.js]: Got it. Updated.

File: assets/js/googlesitekit/modules/components/ModuleSettingsFooter.js [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+	const handleEdit = useCallback( () => {
+		setSettingsDisplayMode( slug, 'edit' );
+	}, [] );
+
+	const handleCancel = useCallback( () => {
+		setSettingsDisplayMode( slug, 'view' );
+	}, [] );
+
+	const handleRemove = useCallback( async () => {
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/googlesitekit/modules/components/ModuleSettingsFooter.js]: Updated.

File: assets/js/googlesitekit/modules/components/ModuleSettingsFooter.js [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+
+	return (
+		<footer className="googlesitekit-settings-module__footer">
+			<div className="mdc-layout-grid">
+				<div className="mdc-layout-grid__inner">
+					<div className="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-8-tablet mdc-layout-grid__cell--span-4-phone">
+						{ buttons }
+					</div>
+					<div className="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-8-tablet mdc-layout-grid__cell--span-4-phone mdc-layout-grid__cell--align-middle mdc-layout-grid__cell--align-right-desktop">
+						{ isEditing && ! autoActivate && canDisconnect && (
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/googlesitekit/modules/components/ModuleSettingsFooter.js]: Updated.

File: assets/js/googlesitekit/modules/components/ModuleSettingsFooter.js [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+				</Link>
+			);
+		} else {
+			buttons.push(
+				<Button key="close-btn" onClick={ () => handleCancel() }>
+					{ __( 'Close', 'google-site-kit' ) }
+				</Button>
+			);
+		}
+	} else if ( allowEdit || ! autoActivate ) {
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/googlesitekit/modules/components/ModuleSettingsFooter.js]: Updated.

File: includes/Core/Modules/Modules.php (hunk 7 above)
Comment (eugene-manuilov) [reply includes/Core/Modules/Modules.php]: Ok, updated.

File: assets/js/googlesitekit/modules/components/ModuleSettings.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+ */
+import Data from 'googlesitekit-data';
+import { STORE_NAME } from '../datastore/constants';
+import ModuleSettingsOverlay from './ModuleSettingsOverlay';
+const { useSelect } = Data;
+
+function ModuleSettings( { children, error, slug } ) {
+	const mode = useSelect( ( select ) => select( STORE_NAME ).getSettingsDisplayMode( slug ) );
+	const isEditing = mode === 'edit';
+	const overlay = mode === 'locked' ? <ModuleSettingsOverlay compress /> : null;
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/googlesitekit/modules/components/ModuleSettings.js]: Added.

**Comment by eugene-manuilov** [comment]
> This PR goes beyond the scope of the issue in that it actually migrates the existing SettingsApp to rely on the new components, which was not mentioned there. If we did that, it would mean we would need to migrate all module's settings first before we could release any of it, which isn't something we should depend on.

@felixarntz changes in SettignsApp and relative components are fully backward compatible, if I don't miss something. So, we don't need to migrate settings for all modules to be able to release it.

File: assets/js/googlesitekit/modules/datastore/modules.js (hunk 5 above)
Comment (eugene-manuilov) [reply assets/js/googlesitekit/modules/datastore/modules.js]: Updated.

**Comment by felixarntz** [comment]
@aaemnnosttv Despite of what I mentioned earlier, looking at this again, maybe we _can_ actually pull it off to replace the existing components right away, based on the improvements added by @eugene-manuilov. Maybe you can have a look at the current PR and see what you think on its state?

One of the things we need to double-check thoroughly is any functionality based on Site Kit module fields which differ between `_googlesitekitLegacyData.modules` and the module data exposed via the API and the `core/modules` store. I'd like to keep the REST API layer simple and we shouldn't add too much to it. For example, the `provides` field is super specific and should be handled on a component level in JS rather than being provided in PHP. Yet, it of course needs to continue to work, so we may need some extra props in JS. Overall, I'd rather implement such minor features by tweaking JS components than by modifying REST API responses or updating datastores because those would be code areas considered "more public".

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

**Pull request closed without merging**